cpp/src/arrow/array/array_dict.cc

```diff
@@ -212,106 +213,155 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
+  return MakeArray(std::move(transposed));
+}
 
-  Status Visit(const DataType& type) {
-    return Status::TypeError("Expected an Index Type of Int or UInt");
+template <int IndexByteWidth>
+Result<std::shared_ptr<Array>> CompactTransposeMap(const std::shared_ptr<ArrayData>& data,
+                                                   Type::type index_type_id,
+                                                   arrow::MemoryPool* pool) {
+  const int64_t index_length = data->length;
+  const int64_t dict_length = data->dictionary->length;
+  const bool indices_are_signed = is_signed_integer(index_type_id);
+
+  if (ARROW_PREDICT_FALSE(dict_length == 0)) {
+    // If the dictionary is empty, we can return the original
+    // array because it can't get any more compact.
+    return std::make_shared<DictionaryArray>(data);
   }
-};
-
-Result<std::unique_ptr<Buffer>> CompactTransposeMap(
-    const std::shared_ptr<ArrayData>& data, MemoryPool* pool,
-    std::shared_ptr<Array>& out_compact_dictionary) {
-  if (data->type->id() != Type::DICTIONARY) {
-    return Status::TypeError("Expected dictionary type");
+  if (ARROW_PREDICT_FALSE(index_length == 0)) {
+    ARROW_ASSIGN_OR_RAISE(auto compact_dictionary,
+                          MakeEmptyArray(data->dictionary->type, pool));
+    ARROW_ASSIGN_OR_RAISE(auto transpose_map, AllocateBuffer(0, pool))
+    return TransposeAndMakeArrayNoInline(data, compact_dictionary,
+                                         std::move(transpose_map), pool);
   }
 
-  const auto& dict_type = checked_cast<const DictionaryType&>(*data->type);
-  CompactTransposeMapVisitor visitor{data, pool, nullptr, nullptr};
-  RETURN_NOT_OK(VisitTypeInline(*dict_type.index_type(), &visitor));
+  ARROW_ASSIGN_OR_RAISE(auto dict_used_bitmap_buffer,
+                        AllocateEmptyBitmap(dict_length, pool));
+  auto* dict_used = dict_used_bitmap_buffer->mutable_data();
+  int64_t dict_used_count;
+  if (indices_are_signed) {
+    using IndexCType =
+        typename internal::int_type_from_byte_width<IndexByteWidth, true>::type;
+    ARROW_ASSIGN_OR_RAISE(dict_used_count,
+                          PopulateBitmapOfUsedIndices<IndexCType>(*data, dict_used));
+  } else {
+    using IndexCType =
+        typename internal::int_type_from_byte_width<IndexByteWidth, false>::type;
+    ARROW_ASSIGN_OR_RAISE(dict_used_count,
+                          PopulateBitmapOfUsedIndices<IndexCType>(*data, dict_used));
+  }
+  if (ARROW_PREDICT_FALSE(dict_used_count == dict_length)) {
+    // The dictionary is already compact, so just return here
+    return std::make_shared<DictionaryArray>(data);
+  }
 
-  out_compact_dictionary = visitor.out_compact_dictionary;
-  return std::move(visitor.output_map);
+  using UnsignedIndexCType =
+      typename internal::int_type_from_byte_width<IndexByteWidth, false>::type;
+  DCHECK_LE(dict_used_count, std::numeric_limits<int32_t>::max());
```

Comment:
Hmm, why would that always be the case? Granted, it's highly unlikely to encounter a dictionary with more than 2^31 distinct elements, but nothing in the spec precludes it.

Reply:
We could return a `Status::CapacityError` here instead of only having a debug assertion.